File: src/main/java/com/app/controller/BatchController.java
package com.app.controller;

import com.app.service.EntityService;
import com.app.util.Sanitizer;
import javax.servlet.http.*;
import java.io.*;
import java.sql.SQLException;

public class BatchController extends HttpServlet {
    
    private EntityService entityService;
    private static final int MAX_ITEMS = 100;
    private static final int MAX_PAGES = 50;
    
    // #L01 - Input bounded by Math.min through utility
    public void processItems(HttpServletRequest req, HttpServletResponse resp) throws IOException {
        int count = Sanitizer.bound(Sanitizer.toInt(req.getParameter("count")), MAX_ITEMS);
        for (int i = 0; i < count; i++) {
            resp.getWriter().write("<li>Item " + i + "</li>");
        }
    }
    
    // #L02 - Input bounded by ternary with constant
    public void processPages(HttpServletRequest req, HttpServletResponse resp) throws IOException {
        int pages = Sanitizer.toInt(req.getParameter("pages"));
        int bounded = pages > MAX_PAGES ? MAX_PAGES : pages;
        for (int i = 0; i < bounded; i++) {
            resp.getWriter().write("<div>Page " + i + "</div>");
        }
    }
    
    // #L03 - Input bounded by if-guard
    public void processRecords(HttpServletRequest req, HttpServletResponse resp) throws IOException {
        int records = Sanitizer.toInt(req.getParameter("records"));
        if (records > MAX_ITEMS) records = MAX_ITEMS;
        if (records < 0) records = 0;
        for (int i = 0; i < records; i++) {
            resp.getWriter().write("<tr><td>" + i + "</td></tr>");
        }
    }
    
    // #L04 - Input bounded by loop condition
    public void processEntries(HttpServletRequest req, HttpServletResponse resp) throws IOException {
        int entries = Sanitizer.toInt(req.getParameter("entries"));
        for (int i = 0; i < entries && i < MAX_ITEMS; i++) {
            resp.getWriter().write("<p>Entry " + i + "</p>");
        }
    }
    
    // #L05 - Input bounded by nested Math.min
    public void processNested(HttpServletRequest req, HttpServletResponse resp) throws IOException {
        int outer = Sanitizer.toInt(req.getParameter("outer"));
        int inner = Sanitizer.toInt(req.getParameter("inner"));
        int boundedOuter = Math.min(outer, 10);
        int boundedInner = Math.min(inner, 10);
        for (int i = 0; i < boundedOuter; i++) {
            for (int j = 0; j < boundedInner; j++) {
                resp.getWriter().write("<span>" + i + "," + j + "</span>");
            }
        }
    }
    
    // #L06 - DB count bounded by constant
    public void processDbCount(HttpServletRequest req, HttpServletResponse resp) throws IOException, SQLException {
        int dbCount = entityService.getAllEntityIds().size();
        int bounded = Math.min(dbCount, MAX_ITEMS);
        for (int i = 0; i < bounded; i++) {
            resp.getWriter().write("<li>DB Item " + i + "</li>");
        }
    }
    
    // #L07 - Regex validated numeric input
    public void processValidated(HttpServletRequest req, HttpServletResponse resp) throws IOException {
        String input = req.getParameter("count");
        if (Sanitizer.isNumeric(input)) {
            int count = Math.min(Integer.parseInt(input), MAX_ITEMS);
            for (int i = 0; i < count; i++) {
                resp.getWriter().write("<div>" + i + "</div>");
            }
        }
    }
    
    // #L08 - Extracted numeric from mixed input
    public void processExtracted(HttpServletRequest req, HttpServletResponse resp) throws IOException {
        String input = req.getParameter("mixed");
        String numeric = Sanitizer.extractNumeric(input);
        int count = Math.min(Integer.parseInt(numeric), MAX_ITEMS);
        for (int i = 0; i < count; i++) {
            resp.getWriter().write("<span>" + i + "</span>");
        }
    }
    
    // #L09 - Modulo bounded
    public void processModulo(HttpServletRequest req, HttpServletResponse resp) throws IOException {
        int input = Sanitizer.toInt(req.getParameter("value"));
        int bounded = input % MAX_ITEMS;
        for (int i = 0; i < bounded; i++) {
            resp.getWriter().write("<li>" + i + "</li>");
        }
    }
    
    // #L10 - Bitwise AND bounded
    public void processBitwise(HttpServletRequest req, HttpServletResponse resp) throws IOException {
        int input = Sanitizer.toInt(req.getParameter("value"));
        int bounded = input & 0x3F;
        for (int i = 0; i < bounded; i++) {
            resp.getWriter().write("<span>" + i + "</span>");
        }
    }
}

File: src/main/csharp/Controllers/BatchController.cs
using System;
using System.Text;
using System.Web.Mvc;
using App.Core;
using App.Services;

namespace App.Controllers
{
    public class BatchController : Controller
    {
        private readonly EntityService _entityService;
        private const int MaxItems = 100;
        private const int MaxPages = 50;
        
        public BatchController(EntityService entityService)
        {
            _entityService = entityService;
        }
        
        // #L01 - Input bounded by Math.Min through utility
        public ActionResult ProcessItems(string count)
        {
            int bounded = Sanitizer.Bound(Sanitizer.ToInt(count), MaxItems);
            var sb = new StringBuilder();
            for (int i = 0; i < bounded; i++)
            {
                sb.Append($"<li>Item {i}</li>");
            }
            return Content(sb.ToString(), "text/html");
        }
        
        // #L02 - Input bounded by ternary with constant
        public ActionResult ProcessPages(string pages)
        {
            int input = Sanitizer.ToInt(pages);
            int bounded = input > MaxPages ? MaxPages : input;
            var sb = new StringBuilder();
            for (int i = 0; i < bounded; i++)
            {
                sb.Append($"<div>Page {i}</div>");
            }
            return Content(sb.ToString(), "text/html");
        }
        
        // #L03 - Input bounded by if-guard
        public ActionResult ProcessRecords(string records)
        {
            int input = Sanitizer.ToInt(records);
            if (input > MaxItems) input = MaxItems;
            if (input < 0) input = 0;
            var sb = new StringBuilder();
            for (int i = 0; i < input; i++)
            {
                sb.Append($"<tr><td>{i}</td></tr>");
            }
            return Content(sb.ToString(), "text/html");
        }
        
        // #L04 - Input bounded by loop condition
        public ActionResult ProcessEntries(string entries)
        {
            int input = Sanitizer.ToInt(entries);
            var sb = new StringBuilder();
            for (int i = 0; i < input && i < MaxItems; i++)
            {
                sb.Append($"<p>Entry {i}</p>");
            }
            return Content(sb.ToString(), "text/html");
        }
        
        // #L05 - Input bounded by nested Math.Min
        public ActionResult ProcessNested(string outer, string inner)
        {
            int outerVal = Math.Min(Sanitizer.ToInt(outer), 10);
            int innerVal = Math.Min(Sanitizer.ToInt(inner), 10);
            var sb = new StringBuilder();
            for (int i = 0; i < outerVal; i++)
            {
                for (int j = 0; j < innerVal; j++)
                {
                    sb.Append($"<span>{i},{j}</span>");
                }
            }
            return Content(sb.ToString(), "text/html");
        }
        
        // #L06 - DB count bounded by constant
        public ActionResult ProcessDbCount()
        {
            int dbCount = _entityService.GetAllEntityIds().Count;
            int bounded = Math.Min(dbCount, MaxItems);
            var sb = new StringBuilder();
            for (int i = 0; i < bounded; i++)
            {
                sb.Append($"<li>DB Item {i}</li>");
            }
            return Content(sb.ToString(), "text/html");
        }
        
        // #L07 - Regex validated numeric input
        public ActionResult ProcessValidated(string count)
        {
            var sb = new StringBuilder();
            if (Sanitizer.IsNumeric(count))
            {
                int bounded = Math.Min(int.Parse(count), MaxItems);
                for (int i = 0; i < bounded; i++)
                {
                    sb.Append($"<div>{i}</div>");
                }
            }
            return Content(sb.ToString(), "text/html");
        }
        
        // #L08 - Extracted numeric from mixed input
        public ActionResult ProcessExtracted(string mixed)
        {
            string numeric = Sanitizer.ExtractNumeric(mixed);
            int bounded = Math.Min(int.Parse(numeric), MaxItems);
            var sb = new StringBuilder();
            for (int i = 0; i < bounded; i++)
            {
                sb.Append($"<span>{i}</span>");
            }
            return Content(sb.ToString(), "text/html");
        }
        
        // #L09 - Modulo bounded
        public ActionResult ProcessModulo(string value)
        {
            int input = Sanitizer.ToInt(value);
            int bounded = input % MaxItems;
            var sb = new StringBuilder();
            for (int i = 0; i < bounded; i++)
            {
                sb.Append($"<li>{i}</li>");
            }
            return Content(sb.ToString(), "text/html");
        }
        
        // #L10 - Bitwise AND bounded
        public ActionResult ProcessBitwise(string value)
        {
            int input = Sanitizer.ToInt(value);
            int bounded = input & 0x3F;
            var sb = new StringBuilder();
            for (int i = 0; i < bounded; i++)
            {
                sb.Append($"<span>{i}</span>");
            }
            return Content(sb.ToString(), "text/html");
        }
    }
}

